README.md
# G-Image
<p align="center">
  <img width="300" height="115" src="images/g-image-logo.png">
</p>
<p align="center">
A cross-platform LabVIEW library for loading, saving, and processing images.
</p>
<p align="center">
<a href="https://www.vipm.io/package/dataflow_g_lib_g_image/"> <img src="https://www.vipm.io/package/dataflow_g_lib_g_image/badge.svg?metric=installs"></a> <a href="https://www.vipm.io/package/dataflow_g_lib_g_image/"><img src="https://www.vipm.io/package/dataflow_g_lib_g_image/badge.svg?metric=stars"></a>
</p>

<p align="center">
    <a href="#whats-new">What's New?</a> -
    <a href="#features">Features</a> -
    <a href="#installation">Installation</a> -
    <a href="#usage">Usage</a> -
	<a href="#building">Building</a> -
	<a href="#comparison">Comparison</a> - 
	<a href="#license">License</a> - 
	<a href="#acknowledgments">Acknowledgments</a>
</p>

## <a id="whats-new"></a>What's New?
* Support for reading PCX images
* Save GIF and animated GIF to memory (issue #3)
* Support GIF and animated GIF transparency (issue #6)
    * Transparency taken from LabVIEW image mask or alpha channel
* Set animated GIF loop count (issue #7)

##### Changed VIs
* _Indexed Color To True Color.vi_ renamed to _To True Color.vi_
    * VI supports converting 24-bit image + mask input to 32-bit ARGB output
* _True Color To Indexed Color.vi_ renamed to _To Indexed Color.vi_
* Removed `Dither` option from _Save GIF Image.vi_ and _Save Animated GIF Image.vi_
    * Image will always be dithered

## <a id="features"></a>Features
* Load images from file or from memory in a range of formats, including:
    * PNG, JPEG, BMP, GIF, Animated GIF, TGA, SVG, PSD, HDR, PIC, PNM, QOI, PCX
* Save images to file or to memory in a range of formats, including:
    * PNG, JPEG, BMP, GIF, Animated GIF, TGA, QOI
* Basic image processing functions:
    * Resize
    * Rotate
    * Flip
    * Quantize (True Color to Indexed Color)

## <a id="installation"></a>Installation
G-Image is published on [vipm.io](https://www.vipm.io/package/dataflow_g_lib_g_image/), and can be installed using VI Package Manager (VIPM). The packages are also available as [github releases](https://github.com/dataflowg/g-image/releases) and can be installed manually using VIPM.

## <a id="usage"></a>Usage
See the example VIs in [Examples](src/LabVIEW/G-Image/Examples) to load, save, and process images.

Unit tests are included and can be run individually, or with the [AST Unit Tester](https://www.autosofttech.net/documents/ast-unit-tester).

## <a id="building"></a>Building
TBD

## <a id="comparison"></a>Comparison
### Image Loading
In addition to loading the formats below from a file, G-Image supports loading images directly from memory.

Format          | G-Image             | LabVIEW Picture Functions
----------------|---------------------|--------------------
PNG             | :heavy_check_mark:¹ | :heavy_check_mark:
JPEG            | :heavy_check_mark:  | :heavy_check_mark:
BMP             | :heavy_check_mark:¹ | :heavy_check_mark:³
GIF             | :heavy_check_mark:¹ | :x:
Animated GIF    | :heavy_check_mark:¹ | :x:
TGA             | :heavy_check_mark:  | :x:
SVG             | :heavy_check_mark:² | :x:
PSD             | :heavy_check_mark:  | :x:
HDR             | :heavy_check_mark:  | :x:
PIC             | :heavy_check_mark:  | :x:
PNM             | :heavy_check_mark:  | :x:
QOI             | :heavy_check_mark:  | :x:
PCX             | :heavy_check_mark:¹ | :x:

¹ *Indexed color images (8-bit and below) are supported, but are converted to 24-bit or 32-bit images. This may change in future updates.*

² *Only vector paths are supported.*

³ *32-bit images are unsupported.*

### Image Saving
In addition to saving the formats below to a file, G-Image supports saving images directly in memory.

Format          | G-Image             | LabVIEW Picture Functions
----------------|---------------------|--------------------
PNG             | :heavy_check_mark:¹ | :heavy_check_mark:
JPEG            | :heavy_check_mark:  | :heavy_check_mark:
BMP             | :heavy_check_mark:¹ | :heavy_check_mark:
GIF             | :heavy_check_mark:  | :heavy_check_mark:²
Animated GIF    | :heavy_check_mark:  | :x:
TGA             | :heavy_check_mark:  | :x:
QOI             | :heavy_check_mark:  | :x:

¹ *Indexed color images (8-bit and below) are supported, but are converted to 24-bit or 32-bit images. This may change in future updates.*

² *The GIF save function isn't in the palette, but can be found in `<vi.lib>\picture\gif.llb\Write GIF File.vi`. Only 8-bit images are supported, and are not compressed with LZW compression.*

## <a id="license"></a>License
This library is built using public domain image libraries where possible. As such, this library is also made available in the public domain. See [LICENSE](LICENSE) for details.

Not all libraries used by G-Image are public domain, but are licensed using open source permissive licenses.

## <a id="acknowledgments"></a>Acknowledgments
This library uses the following open source libraries. Massive thanks to these authors.

Library | Author | License
--------|--------|----------------------
[stb_image.h](https://github.com/nothings/stb) | Sean Barrett | [Unlicense / MIT](https://github.com/nothings/stb/blob/master/LICENSE)
[stb_image_write.h](https://github.com/nothings/stb) | Sean Barrett | [Unlicense / MIT](https://github.com/nothings/stb/blob/master/LICENSE)
[stb_image_resize.h](https://github.com/nothings/stb) | Sean Barrett | [Unlicense / MIT](https://github.com/nothings/stb/blob/master/LICENSE)
[miniz](https://code.google.com/archive/p/miniz)¹ | Rich Geldreich | [Unlicense](https://code.google.com/archive/p/miniz)
[gif-h](https://github.com/charlietangora/gif-h) | Charlie Tangora | [Public Domain](https://github.com/charlietangora/gif-h/blob/master/gif.h)
[nanosvg](https://github.com/memononen/nanosvg) | Mikko Mononen | [zlib](https://github.com/memononen/nanosvg/blob/master/LICENSE.txt)
[qoi](https://github.com/phoboslab/qoi) | Dominic Szablewski | [MIT](https://github.com/phoboslab/qoi/blob/master/LICENSE)
[dr_pcx](https://github.com/mackron/dr_pcx) | David Reid | [Public Domain](https://github.com/mackron/dr_pcx/blob/master/dr_pcx.h)
[msf_gif](https://github.com/notnullnotvoid/msf_gif) | notnullnotvoid | [Public Domain](https://github.com/notnullnotvoid/msf_gif/blob/master/msf_gif.h)

¹ *The linked version is public domain, and is the version used in G-Image. Newer versions of miniz are MIT licensed.*


HISTORY.md
# G-Image Release History
### v0.5.0 - 2023-08-03
* Support for reading PCX images
* Save GIF and animated GIF to memory (issue #3)
* Support GIF and animated GIF transparency (issue #6)
    * Transparency taken from LabVIEW image mask or alpha channel
* Set animated GIF loop count (issue #7)

##### Changed VIs
* _Indexed Color To True Color.vi_ renamed to _To True Color.vi_
    * VI supports converting 24-bit image + mask input to 32-bit ARGB output
* _True Color To Indexed Color.vi_ renamed to _To Indexed Color.vi_
* Removed `Dither` option from _Save GIF Image.vi_ and _Save Animated GIF Image.vi_
    * Image will always be dithered

### v0.4.0 - 2023-05-05
* Support for reading and writing QOI ('Quite OK Image Format') images
    * Lossless 24-bit and 32-bit image type with fast encode + decode
* Support for loading 16-bit per channel images
* Fix incorrect error output when saving GIF files
* Fix incorrect alpha channel when converting certain 8-bit images to 32-bit
* Fix incorrect default scale when loading SVG from UTF-8 path
* Misc. VI doc updates
* Update stb_image.h to v2.28

##### New VIs
* _Save QOI Image.vi_
* _Load Image (16-bit).vi_
* _Convert Image To 8-bit Per Channel.vim_
* _Convert Image To 16-bit Per Channel.vim_
* _Create Mask By Index.vi_

##### Changed VIs
* _Resize Image.vi_ is now polymorphic, accepting either width x height or a scale multiplier

### v0.3.0 - 2023-02-13
* Basic support for loading Scalable Vector Graphics (SVG)
    * Path based vectors only
* Image rotation performance improvement
* Fix when loading certain 8-bit images
* Fix crash when saving images to memory

### v0.2.0 - 2022-08-08
* Support saving GIF and animated GIF (with optional dithering)
* Add PNG compression option when saving
* Add TGA compression option when saving
* Image resize filter option
    * Box, Triangle, Cubic B-Spline, Catmull-Rom, Mitchell
* Improved PNG compression
* New image processing functions
    * Flip Image
    * True Color <-> Indexed Color (with optional dithering)
* Fix image processing on certain 8-bit images

### v0.1.0 - 2022-07-28
* Initial release


src/C++/g_image.h
/*
G-Image - A LabVIEW image library.

v0.4.0

Author - Dataflow_G
GitHub - https://github.com/dataflowg/g-image
Twitter - https://twitter.com/Dataflow_G

/////////////
// HISTORY //
/////////////
v0.2.0
- PNG, TGA compression options
- Image resize options
- Save GIF, Animated GIF
- True Color to Indexed Color function

v0.1.0
- Initial release

/////////////
// LICENSE //
/////////////
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
*/

#ifndef _G_IMAGE_H_
#define _G_IMAGE_H_

#include <stdint.h>
#include <string.h>

#include "miniz.h"

#define STBI_WINDOWS_UTF8
#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"

unsigned char* compress_for_stbiw(unsigned char *data, int data_len, int *out_len, int quality)
{
	uLongf bufSize = compressBound(data_len);
	// note that buf will be free'd by stb_image_write.h
	// with STBIW_FREE() (plain free() by default)
	unsigned char* buf = (unsigned char*)malloc(bufSize);
	if (buf == NULL)
	{
		return NULL;
	}

	if (compress2(buf, &bufSize, data, data_len, quality) != Z_OK)
	{
		free(buf);
		return NULL;
	}
	*out_len = bufSize;

	return buf;
}
#define STBIW_ZLIB_COMPRESS compress_for_stbiw
#if defined( _WIN32 )
#define __STDC_LIB_EXT1__
#endif
#define STBIW_WINDOWS_UTF8
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"

#define STB_IMAGE_RESIZE_IMPLEMENTATION
#include "stb_image_resize.h"

#include "gif.h"

#define MSF_GIF_IMPL
#include "msf_gif.h"

#define NANOSVG_IMPLEMENTATION
#include "nanosvg.h"

#define NANOSVGRAST_IMPLEMENTATION
#include "nanosvgrast.h"

#define QOI_IMPLEMENTATION
#include "qoi.h"

#define DR_PCX_IMPLEMENTATION
#include "dr_pcx.h"

#if defined(_WIN32)
#define LV_DLL_IMPORT  __declspec(dllimport)
#define LV_DLL_EXPORT  __declspec(dllexport)
#define LV_DLL_PRIVATE static
#else
#if defined(__GNUC__) && __GNUC__ >= 4
#define LV_DLL_IMPORT  __attribute__((visibility("default")))
#define LV_DLL_EXPORT  __attribute__((visibility("default")))
#define LV_DLL_PRIVATE __attribute__((visibility("hidden")))
#else
#define LV_DLL_IMPORT
#define LV_DLL_EXPORT
#define LV_DLL_PRIVATE static
#endif
#endif

#define GI_PI           3.14159265358979323846
#define GI_DEG_TO_RAD   0.01745329251994329577
#define GI_RAD_TO_DEG  57.29577951308232087685

typedef int32_t gi_result;

#define GI_SUCCESS				0
// ERRORS
#define GI_E_GENERIC			-1		// Generic error
#define GI_E_MEMORY				-2		// Memory allocation / deallocation error
#define GI_E_UNSUPPORTED		-3		// Unsupported image type
#define GI_E_FILE				-4		// Generic file IO error

// NOTE: This struct is replicated as a cluster in LabVIEW.
// Be careful of alignment issues - LV 32-bit is byte aligned, LV 64-bit is naturally aligned.
typedef struct
{
	void* data;
	uint32_t* colors;
	int32_t* delays;
	int32_t data_size;
	int32_t color_size;
	int32_t width;
	int32_t height;
	int32_t channels;
	int32_t bits_per_channel;
	int32_t layers;
} gi_image_t;

typedef enum
{
	format_save_png,
	format_save_jpg,
	format_save_bmp,
	format_save_gif,
	format_save_tga,
	format_save_qoi,
	format_save_count
} image_format_save_t;

typedef enum
{
	dither_none,
	dither_floyd_steinberg
} dither_type_t;

typedef struct
{
	uint8_t* image_data;
	int32_t image_data_count;
	int32_t image_data_size;
} save_callback_data_t;

////////////////////////////
// LabVIEW CLFN Callbacks //
////////////////////////////
extern "C" LV_DLL_EXPORT int32_t clfn_abort(void* data);

///////////////////////
// LabVIEW image API //
///////////////////////
extern "C" LV_DLL_EXPORT gi_result load_image_from_file(const char* file_name, intptr_t* image_out);
extern "C" LV_DLL_EXPORT gi_result load_image_from_memory(const uint8_t* encoded_image, int32_t encoded_image_size, intptr_t* image_out);
extern "C" LV_DLL_EXPORT gi_result load_image_from_file_16(const char* file_name, intptr_t* image_out);
extern "C" LV_DLL_EXPORT gi_result load_image_from_memory_16(const uint8_t* encoded_image, int32_t encoded_image_size, intptr_t* image_out);
extern "C" LV_DLL_EXPORT gi_result free_image(intptr_t image_ptr);
gi_result free_image(gi_image_t* image);

extern "C" LV_DLL_EXPORT gi_result load_svg_from_file(const char* file_name, float scale, intptr_t* image_out);
extern "C" LV_DLL_EXPORT gi_result load_svg_from_memory(const uint8_t* encoded_image, int32_t encoded_image_count, float scale, intptr_t* image_out);

extern "C" LV_DLL_EXPORT gi_result save_image_to_file(const char* file_name, int32_t format, int32_t width, int32_t height, int32_t channels, const uint8_t* image_data, void* option);
extern "C" LV_DLL_EXPORT gi_result save_image_to_memory(int32_t format, int32_t width, int32_t height, int32_t channels, const uint8_t* image_data, void* option, intptr_t* image_data_out, int32_t* image_data_count_out);
extern "C" LV_DLL_EXPORT gi_result free_data(intptr_t data_ptr);

extern "C" LV_DLL_EXPORT gi_result resize_image(const uint8_t * image_data_in, int32_t width_in, int32_t height_in, int32_t channels_in,
												int32_t width_resize, int32_t height_resize, uint8_t* image_data_out, int32_t filter);
extern "C" LV_DLL_EXPORT gi_result rotate_image(const uint8_t* image_data_in, int32_t width_in, int32_t height_in, int32_t channels_in, int32_t row_stride,
												double angle, uint8_t* image_data_out);

extern "C" LV_DLL_EXPORT gi_result true_color_to_indexed(const uint8_t* image_data_in, int32_t width_in, int32_t height_in, int32_t channels_in, int32_t depth, uint32_t dither, uint8_t* image_data_out, uint32_t* colors_out);

extern "C" LV_DLL_EXPORT gi_result open_write_gif_file(const char* file_name, int32_t width, int32_t height, int32_t depth, int32_t loop_count, intptr_t* writer_ptr);
extern "C" LV_DLL_EXPORT gi_result write_gif_frame_file(intptr_t writer_ptr, const uint8_t* image_data, int32_t width, int32_t height, int32_t delay, uint32_t dither);
extern "C" LV_DLL_EXPORT gi_result close_gif_file(intptr_t writer_ptr);

extern "C" LV_DLL_EXPORT gi_result open_write_gif_memory(int32_t width, int32_t height, int32_t depth, int32_t loop_count, intptr_t* writer_ptr);
extern "C" LV_DLL_EXPORT gi_result write_gif_frame_memory(intptr_t writer_ptr, const uint8_t* image_data, int32_t width, int32_t height, int32_t delay, uint32_t dither);
extern "C" LV_DLL_EXPORT gi_result close_gif_memory(intptr_t writer_ptr, intptr_t* image_data_out, int32_t* image_data_count_out);

gi_result gi_read_image_from_file(const char* file_name, gi_image_t* image);
gi_result gi_read_qoi_from_file(const char* file_name, gi_image_t* image);
bool gi_is_qoi_file(const char* file_name);
gi_result gi_read_image_from_memory(const uint8_t* encoded_image, int32_t encoded_image_count, gi_image_t* image);
gi_result gi_read_qoi_from_memory(const uint8_t* encoded_image, int32_t encoded_image_count, gi_image_t* image);
bool gi_is_qoi_memory(const uint8_t* encoded_image, int32_t encoded_image_count);

gi_result gi_read_pcx_from_file(const char* file_name, gi_image_t* image);
bool gi_is_pcx_file(const char* file_name);
gi_result gi_read_pcx_from_memory(const uint8_t* encoded_image, int32_t encoded_image_count, gi_image_t* image);
bool gi_is_pcx_memory(const uint8_t* encoded_image, int32_t encoded_image_count);

gi_result gi_read_image_from_file_16(const char* file_name, gi_image_t* image);
gi_result gi_read_image_from_memory_16(const uint8_t* encoded_image, int32_t encoded_image_count, gi_image_t* image);

int32_t gi_write_qoi(const char* file_name, int32_t width, int32_t height, int32_t channels, const uint8_t* image_data);
int32_t gi_save_qoi_to_memory(int32_t width, int32_t height, int32_t channels, const uint8_t* image_data, save_callback_data_t* callback_data);
int32_t gi_write_gif(const char* file_name, int32_t width, int32_t height, const uint8_t* image_data, dither_type_t dither);
int32_t gi_save_gif_to_memory(int32_t width, int32_t height, const uint8_t* image_data, save_callback_data_t* callback_data);
//////////////////////////
// Ancilliary Functions //
//////////////////////////
STBIDEF unsigned char *stbi_load_extended(stbi__context* s, int* x, int* y, int* n, int* layers, int** delays)
{
	unsigned char *result = 0;

	if (stbi__gif_test(s))
		return (uint8_t*)stbi__load_gif_main(s, delays, x, y, layers, n, 4);

	stbi__result_info ri;
	result = (uint8_t*)stbi__load_main(s, x, y, n, 4, &ri, 8);
	*layers = !!result;
	*delays = (int*)calloc(1, sizeof(int));

	if (ri.bits_per_channel != 8) {
		STBI_ASSERT(ri.bits_per_channel == 16);
		result = stbi__convert_16_to_8((stbi__uint16 *)result, *x, *y, 4);
		ri.bits_per_channel = 8;
	}

	return result;
}

STBIDEF unsigned char *stbi_load_from_memory_extended(const unsigned char* buffer, int len, int* x, int* y, int* n, int* layers, int** delays)
{
	stbi__context s;
	stbi__start_mem(&s, buffer, len);
	return stbi_load_extended(&s, x, y, n, layers, delays);
}

STBIDEF unsigned char *stbi_load_from_file_extended(const char* filename, int* x, int* y, int* n, int* layers, int** delays)
{
	FILE *f;
	stbi__context s;
	unsigned char *result = 0;

	if (!(f = stbi__fopen(filename, "rb")))
		return stbi__errpuc("can't fopen", "Unable to open file");

	stbi__start_file(&s, f);
	result = stbi_load_extended(&s, x, y, n, layers, delays);
	fclose(f);

	return result;
}

// Creates a gif file.
// The input GIFWriter is assumed to be uninitialized.
// The delay value is the time between frames in hundredths of a second - note that not all viewers pay much attention to this value.
bool GifBeginUtf8(GifWriter* writer, const char* filename, uint32_t width, uint32_t height, uint32_t delay, int32_t bitDepth = 8, bool dither = false)
{
	(void)bitDepth; (void)dither; // Mute "Unused argument" warnings

	writer->f = stbiw__fopen(filename, "wb");
	if (!writer->f) return false;

	writer->firstFrame = true;

	// allocate
	writer->oldImage = (uint8_t*)GIF_MALLOC(width * height * 4);

	fputs("GIF89a", writer->f);

	// screen descriptor
	fputc(width & 0xff, writer->f);
	fputc((width >> 8) & 0xff, writer->f);
	fputc(height & 0xff, writer->f);
	fputc((height >> 8) & 0xff, writer->f);

	fputc(0xf0, writer->f);  // there is an unsorted global color table of 2 entries
	fputc(0, writer->f);     // background color
	fputc(0, writer->f);     // pixels are square (we need to specify this because it's 1989)

	// now the "global" palette (really just a dummy palette)
	// color 0: black
	fputc(0, writer->f);
	fputc(0, writer->f);
	fputc(0, writer->f);
	// color 1: also black
	fputc(0, writer->f);
	fputc(0, writer->f);
	fputc(0, writer->f);

	if (delay != 0)
	{
		// animation header
		fputc(0x21, writer->f); // extension
		fputc(0xff, writer->f); // application specific
		fputc(11, writer->f); // length 11
		fputs("NETSCAPE2.0", writer->f); // yes, really
		fputc(3, writer->f); // 3 bytes of NETSCAPE2.0 data

		fputc(1, writer->f); // JUST BECAUSE
		fputc(0, writer->f); // loop infinitely (byte 0)
		fputc(0, writer->f); // loop infinitely (byte 1)

		fputc(0, writer->f); // block terminator
	}

	return true;
}

NSVGimage* nsvgParseFromFileUtf8(const char* filename, const char* units, float dpi)
{
	FILE* fp = NULL;
	size_t size;
	char* data = NULL;
	NSVGimage* image = NULL;

	fp = stbi__fopen(filename, "rb");
	if (!fp) goto error;
	fseek(fp, 0, SEEK_END);
	size = ftell(fp);
	fseek(fp, 0, SEEK_SET);
	data = (char*)malloc(size + 1);
	if (data == NULL) goto error;
	if (fread(data, 1, size, fp) != size) goto error;
	data[size] = '\0';	// Must be null terminated.
	fclose(fp);
	image = nsvgParse(data, units, dpi);
	free(data);

	return image;

error:
	if (fp) fclose(fp);
	if (data) free(data);
	if (image) nsvgDelete(image);
	return NULL;
}

NSVGimage* nsvgParseFromMemory(const uint8_t* encoded_image, int32_t encoded_image_count, const char* units, float dpi)
{
	size_t size = encoded_image_count;
	char* data = NULL;
	NSVGimage* image = NULL;

	data = (char*)malloc(size + 1);
	if (data == NULL) goto error;
	memcpy(data, encoded_image, size);
	data[size] = '\0';	// Must be null terminated.
	image = nsvgParse(data, units, dpi);
	free(data);

	return image;

error:
	if (data) free(data);
	if (image) nsvgDelete(image);
	return NULL;
}

int qoi_write_utf8(const char *filename, const void *data, const qoi_desc *desc)
{
	FILE *f = stbiw__fopen(filename, "wb");
	int size;
	void *encoded;

	if (!f)
	{
		return 0;
	}

	encoded = qoi_encode(data, desc, &size);
	if (!encoded)
	{
		fclose(f);
		return 0;
	}

	fwrite(encoded, 1, size, f);
	fclose(f);

	QOI_FREE(encoded);
	return size;
}

void *qoi_read_utf8(const char *filename, qoi_desc *desc, int channels)
{
	FILE *f = stbi__fopen(filename, "rb");
	int size, bytes_read;
	void *pixels, *data;

	if (!f)
	{
		return NULL;
	}

	fseek(f, 0, SEEK_END);
	size = ftell(f);
	if (size <= 0)
	{
		fclose(f);
		return NULL;
	}
	fseek(f, 0, SEEK_SET);

	data = QOI_MALLOC(size);
	if (!data)
	{
		fclose(f);
		return NULL;
	}

	bytes_read = fread(data, 1, size, f);
	fclose(f);

	pixels = qoi_decode(data, bytes_read, desc, channels);
	QOI_FREE(data);
	return pixels;
}

drpcx_uint8* drpcx_load_file_utf8(const char* filename, drpcx_bool32 flipped, int* x, int* y, int* internalComponents, int desiredComponents)
{
	FILE* pFile = stbi__fopen(filename, "rb");

	if (pFile == NULL)
	{
		return NULL;
	}

	drpcx_uint8* pImageData = drpcx_load(drpcx__on_read_stdio, pFile, flipped, x, y, internalComponents, desiredComponents);

	fclose(pFile);
	return pImageData;
}

#endif
